drivers/ethernet/eth_lan9250.c

```diff
@@ -0,0 +1,769 @@
+	uint8_t instr[2] = {(LAN9250_TX_DATA_FIFO >> 8) & 0xFF, (LAN9250_TX_DATA_FIFO & 0xFF)};
+
+	struct spi_buf tx_buf[3];
+
+	tx_buf[0].buf = &cmd;
+	tx_buf[0].len = ARRAY_SIZE(cmd);
+
+	tx_buf[1].buf = &instr;
+	tx_buf[1].len = ARRAY_SIZE(instr);
+
+	tx_buf[2].buf = data_buffer;
+	tx_buf[2].len = buf_len;
+
+	struct spi_buf_set tx = {.buffers = tx_buf, .count = 3};
+
+	ret |= spi_transceive_dt(&config->spi, &tx, NULL);
+
+	return ret;
+}
+
+static int lan9250_read_buf(const struct device *dev, uint8_t *data_buffer, uint16_t buf_len)
+{
+	const struct lan9250_config *config = dev->config;
+	uint8_t cmd[1] = {LAN9250_SPI_INSTR_READ};
+	uint8_t instr[2] = {(LAN9250_RX_DATA_FIFO >> 8) & 0xFF, (LAN9250_RX_DATA_FIFO & 0xFF)};
+	int ret = 0;
+
+	struct spi_buf tx_buf[3];
+
+	tx_buf[0].buf = &cmd;
+	tx_buf[0].len = ARRAY_SIZE(cmd);
+	tx_buf[1].buf = &instr;
+	tx_buf[1].len = ARRAY_SIZE(instr);
+	tx_buf[2].buf = NULL;
+	tx_buf[2].len = buf_len;
+
+	struct spi_buf_set tx = {.buffers = tx_buf, .count = 3};
+	struct spi_buf rx_buf[3];
+
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = 1;
+	rx_buf[1].buf = NULL;
+	rx_buf[1].len = 2;
+	rx_buf[2].buf = data_buffer;
+	rx_buf[2].len = buf_len;
+
+	const struct spi_buf_set rx = {.buffers = rx_buf, .count = 3};
+
+	ret |= spi_transceive_dt(&config->spi, &tx, &rx);
+
+	return ret;
+}
+
+static int lan9250_rx(const struct device *dev)
+{
+	const struct lan9250_config *config = dev->config;
+	struct lan9250_runtime *ctx = dev->data;
+	uint8_t pktcnt;
+	uint32_t tmp;
+	uint16_t pkt_len;
+	uint16_t lengthfr;
+
+	/* Check valid packet count in RX FIFO */
+	lan9250_read_sys_reg(dev, LAN9250_RX_FIFO_INF, &tmp);
+	pktcnt = (tmp & LAN9250_RX_FIFO_INF_RXSUSED) >> 16;
+
+	if (!pktcnt) {
+		return 0;
+	}
+
+	k_sem_take(&ctx->tx_rx_sem, K_FOREVER);
+	lan9250_write_sys_reg(dev, LAN9250_INT_EN, 0);
+
+	while (pktcnt--) {
+
+		/* Get the length of the packet */
+		lan9250_read_sys_reg(dev, LAN9250_RX_STATUS_FIFO, &tmp);
+
+		pkt_len = (tmp & LAN9250_RX_STS_PACKET_LEN) >> 16;
+		pkt_len -= 4;
+		lengthfr = pkt_len;
+
+		struct net_buf *pkt_buf;
+		struct net_pkt *pkt;
+
+		/* Get the frame from the buffer */
+		pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, pkt_len, AF_UNSPEC, 0,
+						   K_MSEC(config->timeout));
+		if (!pkt) {
+			LOG_ERR("%s: Could not allocate rx buffer", dev->name);
+			eth_stats_update_errors_rx(ctx->iface);
+			return 0;
+		}
+
+		pkt_buf = pkt->buffer;
+
+		lan9250_read_sys_reg(dev, LAN9250_RX_DATA_FIFO, &tmp);
+
+		do {
+
+			size_t frag_len;
+			uint8_t *data_ptr;
+			size_t spi_frame_len;
+
+			data_ptr = pkt_buf->data;
+
+			/* Review the space available for the new frag */
+			frag_len = net_buf_tailroom(pkt_buf);
+
+			if (pkt_len > frag_len) {
+				spi_frame_len = frag_len;
+			} else {
+				spi_frame_len = pkt_len;
+			}
+
+			/* Review the space available for the new frag */
+			lan9250_read_buf(dev, data_ptr, spi_frame_len);
+			net_buf_add(pkt_buf, pkt_len);
+
```

Comment:
I don't think it is needed because it is already checked how much space is available at the end of the buffer with net_buf_tailroom(). Similar to the other SPI Ethernet modules (w5500, enc28j60, eth_enc424j600)